File: packages/client-flutter/lib/src/transport.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import 'events.dart';

class TransportConfig {
  final String? url;
  final String host;
  final int port;
  final Duration reconnectInterval;
  final int maxQueueSize;
  final bool silent;

  const TransportConfig({
    this.url,
    this.host = 'localhost',
    this.port = 9090,
    this.reconnectInterval = const Duration(seconds: 2),
    this.maxQueueSize = 1000,
    this.silent = true,
  });

  Uri get uri {
    if (url != null) return Uri.parse(url!);
    return Uri.parse('ws://$host:$port');
  }
}

enum _ConnState { idle, connecting, open, closed }

class Transport {
  final TransportConfig config;
  WebSocketChannel? _channel;
  _ConnState _state = _ConnState.idle;
  final List<Map<String, dynamic>> _queue = [];
  Timer? _reconnectTimer;
  Timer? _heartbeatTimer;
  HandshakePayload? _handshake;
  void Function(Map<String, dynamic>)? _onMessage;

  Transport(this.config);

  void setHandshake(HandshakePayload payload) {
    _handshake = payload;
  }

  void onMessage(void Function(Map<String, dynamic>) handler) {
    _onMessage = handler;
  }

  Future<void> connect() async {
    if (_state == _ConnState.connecting || _state == _ConnState.open) return;

    _state = _ConnState.connecting;
    _log('connecting to ${config.uri}');

    final WebSocketChannel ch;
    try {
      ch = WebSocketChannel.connect(config.uri);
    } catch (e) {
      _log('connect threw: $e');
      _state = _ConnState.closed;
      _scheduleReconnect();
      return;
    }
    _channel = ch;

    ch.stream.listen(
      (raw) {
        if (_onMessage == null) return;
        try {
          final text = raw is String ? raw : utf8.decode(raw as List<int>);
          final msg = jsonDecode(text) as Map<String, dynamic>;
          _onMessage!(msg);
        } catch (_) {
          /* ignore */
        }
      },
      onDone: _onClose,
      onError: (e) => _log('stream error: $e'),
      cancelOnError: false,
    );

    try {
      await ch.ready;
    } catch (e) {
      _log('ready failed: $e');
      // onDone handler will fire and reconnect for us.
      return;
    }

    _log('connected');
    _onOpen();
  }

  void _log(String message) {
    if (config.silent) return;
    debugPrint('[owlscope] $message');
  }

  void _onOpen() {
    if (_state == _ConnState.open) return;
    _state = _ConnState.open;
    if (_handshake != null) {
      _sendRaw({'type': 'handshake', 'payload': _handshake!.toJson()});
    }
    _flushQueue();
    _startHeartbeat();
  }

  void _onClose() {
    _state = _ConnState.closed;
    _stopHeartbeat();
    _channel = null;
    _scheduleReconnect();
  }

  void send(Map<String, dynamic> message) {
    if (_state == _ConnState.open && _channel != null) {
      _sendRaw(message);
    } else {
      _queue.add(message);
      if (_queue.length > config.maxQueueSize) {
        _queue.removeRange(0, _queue.length - config.maxQueueSize);
      }
    }
  }

  void _sendRaw(Map<String, dynamic> message) {
    final ch = _channel;
    if (ch == null) return;
    try {
      ch.sink.add(jsonEncode(message));
    } catch (_) {
      /* ignore */
    }
  }

  void _flushQueue() {
    while (_queue.isNotEmpty && _state == _ConnState.open) {
      final next = _queue.removeAt(0);
      _sendRaw(next);
    }
  }

  void _scheduleReconnect() {
    if (_reconnectTimer != null) return;
    _reconnectTimer = Timer(config.reconnectInterval, () {
      _reconnectTimer = null;
      // Fire-and-forget; reconnect will reschedule itself if it fails again.
      connect();
    });
  }

  void _startHeartbeat() {
    _stopHeartbeat();
    _heartbeatTimer = Timer.periodic(const Duration(seconds: 30), (_) {
      send({'type': 'ping', 'timestamp': DateTime.now().millisecondsSinceEpoch});
    });
  }

  void _stopHeartbeat() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
  }

  Future<void> close() async {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _stopHeartbeat();
    final ch = _channel;
    _channel = null;
    _state = _ConnState.closed;
    try {
      await ch?.sink.close();
    } catch (_) {
      /* ignore */
    }
  }
}

File: packages/client-js/src/core/transport.ts
import type {
  ClientToServerMessage,
  ServerToClientMessage,
  HandshakePayload,
} from '../protocol.js';

// Inlined so the published SDK has no runtime dependency on the
// monorepo-local `../protocol.js` package — only its types.
const DEFAULT_HOST = 'localhost';
const DEFAULT_PORT = 9090;

export interface TransportConfig {
  url?: string;
  host?: string;
  port?: number;
  reconnectIntervalMs?: number;
  maxQueueSize?: number;
  silent?: boolean;
}

type ConnectionState = 'idle' | 'connecting' | 'open' | 'closed';

type WSCtor = new (url: string) => WebSocket;

function resolveWebSocket(): WSCtor | null {
  const g = globalThis as { WebSocket?: WSCtor };
  if (g.WebSocket) return g.WebSocket;
  return null;
}

export class Transport {
  private url: string;
  private reconnectIntervalMs: number;
  private maxQueueSize: number;
  private silent: boolean;

  private ws: WebSocket | null = null;
  private state: ConnectionState = 'idle';
  private queue: ClientToServerMessage[] = [];
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;

  private handshakePayload: HandshakePayload | null = null;
  private onCommand: ((msg: ServerToClientMessage) => void) | null = null;

  constructor(config: TransportConfig = {}) {
    const host = config.host ?? DEFAULT_HOST;
    const port = config.port ?? DEFAULT_PORT;
    this.url = config.url ?? `ws://${host}:${port}`;
    this.reconnectIntervalMs = config.reconnectIntervalMs ?? 2000;
    this.maxQueueSize = config.maxQueueSize ?? 1000;
    this.silent = config.silent ?? true;
  }

  setHandshake(payload: HandshakePayload) {
    this.handshakePayload = payload;
  }

  onMessage(handler: (msg: ServerToClientMessage) => void) {
    this.onCommand = handler;
  }

  connect() {
    if (this.state === 'connecting' || this.state === 'open') return;
    const WS = resolveWebSocket();
    if (!WS) {
      if (!this.silent) {
        // eslint-disable-next-line no-console
        console.warn('[owlscope] WebSocket not available in this environment.');
      }
      return;
    }

    this.state = 'connecting';
    let socket: WebSocket;
    try {
      socket = new WS(this.url);
    } catch (err) {
      this.state = 'closed';
      this.scheduleReconnect();
      return;
    }
    this.ws = socket;

    socket.addEventListener('open', () => {
      this.state = 'open';
      if (this.handshakePayload) {
        this.sendRaw({ type: 'handshake', payload: this.handshakePayload });
      }
      this.flushQueue();
      this.startHeartbeat();
    });

    socket.addEventListener('message', (ev: MessageEvent) => {
      if (!this.onCommand) return;
      try {
        const data = typeof ev.data === 'string' ? ev.data : '';
        if (!data) return;
        const msg = JSON.parse(data) as ServerToClientMessage;
        this.onCommand(msg);
      } catch {
        /* ignore */
      }
    });

    socket.addEventListener('close', () => {
      this.state = 'closed';
      this.stopHeartbeat();
      this.ws = null;
      this.scheduleReconnect();
    });

    socket.addEventListener('error', () => {
      // close handler will run as well
    });
  }

  send(msg: ClientToServerMessage) {
    if (this.state === 'open' && this.ws) {
      this.sendRaw(msg);
    } else {
      this.queue.push(msg);
      if (this.queue.length > this.maxQueueSize) {
        this.queue.splice(0, this.queue.length - this.maxQueueSize);
      }
    }
  }

  private sendRaw(msg: ClientToServerMessage) {
    if (!this.ws) return;
    try {
      this.ws.send(JSON.stringify(msg));
    } catch {
      /* ignore */
    }
  }

  private flushQueue() {
    while (this.queue.length > 0 && this.state === 'open') {
      const msg = this.queue.shift();
      if (msg) this.sendRaw(msg);
    }
  }

  private scheduleReconnect() {
    if (this.reconnectTimer) return;
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.connect();
    }, this.reconnectIntervalMs);
  }

  private startHeartbeat() {
    this.stopHeartbeat();
    this.heartbeatTimer = setInterval(() => {
      this.send({ type: 'ping', timestamp: Date.now() });
    }, 30000);
  }

  private stopHeartbeat() {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
  }

  close() {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this.stopHeartbeat();
    if (this.ws) {
      try {
        this.ws.close();
      } catch {
        /* ignore */
      }
    }
    this.ws = null;
    this.state = 'closed';
  }
}
